File: src/main/java/com/example/ide/psi/CssModulesIndexedStylesVarPsiReferenceContributor.java
package com.example.ide.psi;
import com.example.ide.css.QCssModulesUtil;
import com.intellij.lang.javascript.psi.JSLiteralExpression;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.*;
import com.intellij.psi.css.*;
import com.intellij.util.ProcessingContext;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;


/**
 * Adds a PSI references from an indexed string literal on a styles object to its corresponding class name.
 * For example, the 'normal' in styles['normal'] will point to the '.normal {}' CSS class in a requirement'd stylesheet.
 * If not Founded  , will point the First CssClassSelector in stylesheet
 */
public class CssModulesIndexedStylesVarPsiReferenceContributor extends PsiReferenceContributor {
    private static final String DOT = ".";
    @Override
    public void registerReferenceProviders(@NotNull PsiReferenceRegistrar registrar) {
        QCssModulesUtil.initContainer();
        registrar.registerReferenceProvider(QCssModulesUtil.STRING_PATTERN, new PsiReferenceProvider() {
            @NotNull
            @Override
            public PsiReference @NotNull [] getReferencesByElement(@NotNull PsiElement element, @NotNull ProcessingContext context) {
                final PsiElement cssClassNamesImportOrRequire = QCssModulesUtil.getCssClassNamesImportOrRequireDeclaration((JSLiteralExpression) element);
                if (cssClassNamesImportOrRequire != null) {
                    final String literalClass = DOT + StringUtils.stripStart(StringUtils.stripEnd(element.getText(), "\"'"), "\"'");
                    final Ref<StylesheetFile> referencedStyleSheet = new Ref<>();
                    final PsiElement psiElement = QCssModulesUtil.getCssClass(cssClassNamesImportOrRequire, literalClass, referencedStyleSheet);
                    if (psiElement != null) {
                        return new PsiReference[]{new PsiReferenceBase<PsiElement>(element) {

                            @Override
                            public @NotNull PsiElement resolve() {
                                return psiElement;
                            }

                            @NotNull
                            @Override
                            public Object @NotNull [] getVariants() {
                                return new Object[0];
                            }
                        }};
                    }else {
                       if (referencedStyleSheet.get() != null) {
                           final TextRange rangeInElement = TextRange.from(1, element.getTextLength() - 2); // minus string quotes
                           return new PsiReference[]{new CssModulesUnknownClassPsiReference(element, rangeInElement, referencedStyleSheet.get())};
                       }
                    }
                }
                return PsiReference.EMPTY_ARRAY;
            }
        });
    }

}

File: src/main/kotlin/com/example/ide/psi/CssModulesIndexedStylesVarPsiReferenceContributor.kt
package com.example.ide.psi


import com.example.ide.completion.findReferenceStyleFile
import com.intellij.lang.ecmascript6.psi.ES6ImportedBinding
import com.intellij.lang.javascript.psi.JSLiteralExpression
import com.intellij.lang.javascript.psi.JSReferenceExpression
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.*
import com.intellij.psi.css.StylesheetFile
import com.intellij.psi.filters.ElementFilter
import com.intellij.psi.filters.position.FilterPattern
import com.intellij.util.ProcessingContext
import org.jetbrains.annotations.NotNull


/**
 * Reference provider for styles["className"] syntax
 */
class CssModuleIndexedReferenceProvider : PsiReferenceProvider() {

    override fun getReferencesByElement(
        element: PsiElement,
        context: ProcessingContext
    ): Array<PsiReference> {
        if (element !is JSLiteralExpression) return PsiReference.EMPTY_ARRAY
        val name = element.stringValue?.trim().orEmpty()
        if (name.isBlank()) return PsiReference.EMPTY_ARRAY
        val styleFile = findReferenceStyleFile(element) ?: return PsiReference.EMPTY_ARRAY

        // Always return a dynamic reference that resolves fresh each time
        return arrayOf(CssModuleClassReference(element, styleFile, name))
    }
}

/**
 * Reference provider for styles.className syntax
 */
class CssModuleDotReferenceProvider : PsiReferenceProvider() {

    override fun getReferencesByElement(
        element: PsiElement,
        context: ProcessingContext
    ): Array<PsiReference> {
        if (element !is JSReferenceExpression) return PsiReference.EMPTY_ARRAY

        // Get the qualifier (the part before the dot, e.g., "styles" in "styles.app")
        val qualifier = element.qualifier
        if (qualifier !is JSReferenceExpression) return PsiReference.EMPTY_ARRAY

        // Resolve the qualifier to check if it's a CSS module import
        val resolved = qualifier.reference?.resolve()
        if (resolved !is ES6ImportedBinding) return PsiReference.EMPTY_ARRAY

        val referencedElements = resolved.findReferencedElements()
        if (referencedElements.isEmpty()) return PsiReference.EMPTY_ARRAY

        val styleFile = referencedElements.first() as? StylesheetFile ?: return PsiReference.EMPTY_ARRAY

        // Get the property name (the part after the dot)
        val propertyName = element.referenceName ?: return PsiReference.EMPTY_ARRAY

        return arrayOf(CssModuleClassReference(element, styleFile, propertyName))
    }
}


// Filter for styles.className syntax
private val DOT_ACCESS_FILTER = PlatformPatterns.psiElement(JSReferenceExpression::class.java).and(
    FilterPattern(
        object : ElementFilter {
            override fun isAcceptable(element: Any?, context: PsiElement?): Boolean {

                return element is JSReferenceExpression
                        && element.reference?.resolve() !== null
                        && (element.reference?.resolve() as ES6ImportedBinding).findReferencedElements()
                    .isNotEmpty()
                        && (element.reference?.resolve() as ES6ImportedBinding).findReferencedElements()
                    .first() is StylesheetFile
            }

            override fun isClassAcceptable(hintClass: Class<*>?): Boolean {
                return JSReferenceExpression::class.java.isAssignableFrom(hintClass!!)
            }
        }
    ))

class CssModulesIndexedStylesVarPsiReferenceContributor : PsiReferenceContributor() {
    override fun registerReferenceProviders(@NotNull registrar: PsiReferenceRegistrar) {
        // Register provider for styles.className syntax
        registrar.registerReferenceProvider(
            DOT_ACCESS_FILTER, CssModuleDotReferenceProvider()
        )
    }
}

fun isStyleIndex(element: JSLiteralExpression): Boolean = findReferenceStyleFile(element) !== null